dependencies/Indy10/System/IdStackUnix.pas

```diff
@@ -271,17 +311,92 @@ procedure InitSockAddr_in6(var VSock : SockAddr_in6);
   {$ENDIF}
   VSock.sin6_family := PF_INET6;
 end;
-//
+
+{$IFDEF DNSCACHE}
+
+{ TIdStackNameResolution }
+
+constructor TIdStackNameResolution.Create(AWhen: TDateTime; AHostName: String; AIP: String);
+begin
+  Inherited Create;
+  FWhen := AWhen;
+  FHost := AHostName;
+  FIP := AIP;
+end;
+
+{$ENDIF}
+
+{ TIdStackUnix }
+
 constructor TIdStackUnix.Create;
 begin
   inherited Create;
+  {$IFDEF DNSCACHE}
+  FHostNames := TObjectList.create;
+  FHostNames.OwnsObjects := true;
+  FLock := TIdCriticalSection.Create;
+  {$ENDIF}
 end;
 
 destructor TIdStackUnix.Destroy;
 begin
+  {$IFDEF DNSCACHE}
+  FreeAndNil(FHostNames);
+  FreeAndNil(FLock);
+  {$ENDIF}
   inherited Destroy;
 end;
 
+{$IFDEF DNSCACHE}
+function TIdStackUnix.GetCachedHostAddress(AHostName : string; var VIP : String) : boolean;
+var
+  oldAge : TDateTime;
+  i : integer;
+  t : TIdStackNameResolution;
+begin
+  oldAge := now - (DNS_CACHE_MINUTES_EXPIRY * 0.000694444444444);
+  result := false;
+  VIP := '';
+  FLock.Enter;
+  try
+    for i := FHostNames.count - 1 downto 0 do
+    begin
+      t := FHostNames[i] as TIdStackNameResolution;

```

Comment:
Since all of the objects are TIdStackNameResolution, I would suggest using a hard cast instead of invoking the overhead of the 'as' operator.